### DISCUSSION TITLE
Image Message overlay like blurring

### DISCUSSION THREAD

**Discussion by HyprUsr** [flyerhq/flutter_chat_ui#666] [Ideas & Requests] [answered]
It would be great to extend the Image Message with the option of overlay like blurring (similar to this solution https://stackoverflow.com/a/59154739/4310309), it should be possible to add it to the new implementation in "FlyerChatImageMessageState"
The use case is for when you want to send a spoiler photo (like Telegram) that the other user can't initially see, but they only see if they choose to see. Another use case could be NSFW pictures that you can block first and let the user decide to see.
I can create a PR for it if it is find to have

**Comment by demchenkoalex** [comment]
Hello! Since it is per image basis, we would need a property of the image message class right? I am thinking something like `blurred` (boolean)

**Reply by HyprUsr** [reply]
yes and no, I'd prefer to have it more flexible, for example a property with the type Widget that comes over the image, then anyone can customize it, like add a button over the blurred image so they can unblur the image, ykwim?

**Reply by demchenkoalex** [reply]
yes you could customise anything via `Builder` but it applies to **all** messages. So you can customise how your overlay looks, however you still need a way to toggle something for **each** message. You can't modify state of each message, it is what controller does, and controller operates on message classes.

**Reply by HyprUsr** [reply]
You're right, I'll use the builder 👍🏼

**Reply by demchenkoalex** [reply]
ok so we agree that we need - a) easy UI customisation via builder, b) property on a message to toggle state. Right?

**Reply by HyprUsr** [reply]
I'm actually using `imageMessageBuilder` now and it fixes the problem :)
Maybe my use case is very specific, I'm not sure if it is useful for others as well tbh

**Reply by demchenkoalex** [reply]
interesting :D can I see the code?

**Reply by HyprUsr** [reply]
sure :) here: https://gist.github.com/Jafarili/385c22ac15e889ba2a6cfe719b290645

**Reply by demchenkoalex** [reply, marked as answer]
I improved this so there is no need to redefine the whole image, but rather image message builder will accept `overlay` param like this

```dart
imageMessageBuilder: (context, message) => FlyerChatImageMessage(
  message: message,
  overlay: GestureDetector(
    onTap: () {
      final newMessage = message.copyWith(overlay: false);
      _chatController.update(message, newMessage);
    },
    child: Positioned.fill(
      child: ClipRect(
        child: BackdropFilter(
          filter: ImageFilter.blur(sigmaX: 30, sigmaY: 30),
          child: Container(
            color: Colors.white.withValues(alpha: 0.5),
            child: const Center(
              child: Text('Tap to reveal'),
            ),
          ),
        ),
      ),
    ),
  ),
),
```

and then you create image messages with `overlay: true`. In the example above tap will copy message with `overlay: false` making it disappear.

https://github.com/user-attachments/assets/693c6116-0faf-4119-9018-5c2c18592ad4
